### ISSUE TITLE
A client must identify to repository with unique id

### ISSUE THREAD

**Issue by enikao** [LionWeb-io/specification#241]
With bulk API _ids_ command, we guarantee to a client that some unknown node ids are reserved for this client.
As the repository needs to keep track of reserved ids per client, each client MUST provide a unique client id when connecting to the repository.

The client id MUST be compatible to node id string (#31).
Rationale: sufficient value space for everything, keep things simple.

A client MAY connect more than once to the same repository at the same time with the same client id.
Rationale: We will have multiple connections anyway (e.g. bulk + delta, maybe also language API etc.).

It is out of scope of LionWeb to guarantee the uniqueness of the client id.
Rationale: Lots of overhead for little benefit. The repository would need to have a registry of client ids, and an API to request a new one (and maybe to unregister).
It's sufficiently easy for a client to come up with a probably unique client id (e.g. [UUID version 4](https://en.wikipedia.org/wiki/Universally_unique_identifier)).

 The repository does not apply any uniqueness checks.
Rationale: How could it? If a client can connect several times, how to distinguish one client connecting twice from two clients connecting with the same id?

**Comment by joswarmer** [comment]
Does this mean that every API call gets an extra parameter?

**Comment by enikao** [comment]
For a stateless transport protocol (e.g. HTTP), yes. If there would be an authentication proxy, we might reuse info from there.
In a stateful protocol (e.g. Websockets), the client could send its id when connecting.

**Comment by joswarmer** [comment]
So ... the repository API does get an extra parameter for each request, but in the case of an authentication proxy the parameter will be added to the call from the client.

Is this right?

**Comment by joswarmer** [comment]
Right, so in a stateful protocol the number of parameters of the API methods is different (no clientId parameter) then in a stateless protocol? That feels rather strange.

**Comment by enikao** [comment]
I'd say semantically we always have the same number of parameters. In stateful protocols, one of them is provided implicitly.

**Comment by enikao** [comment]
**Decision: add client id to signon query**

Authentication / Authorization (i.e. "is this communication endpoint allowed to talk in the name of that client id?") is out of scope of LW. It needs to be handled e.g. by a proxy in front of LW repo.
